### PR TITLE
feat: 87% fewer toast messages

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

@annikaschmid 's [nifty analysis](https://us.posthog.com/project/2/notebooks/k7IXXTRs) found we toast - alot! We do this in ways that isn't particularly helpful, primarily when:
1. You aren't connected to the internet (often happening when you close your laptop or some other thing that loses connection)
2. You are logged out for one reason or another and so get 401 errors.

## Changes

* Only show a toast if we have a proper error message - if we think it is offline, show a top level banner instead
* If we see a 401, we call the @me endpoint to see if we are for sure logged out and then actually do the logout (redirecting to login)
* Also refactors the api a bit to have some better standard fetch handling

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: frontend/src/lib/api.ts [hunk 1]
```diff
@@ -2119,10 +2089,36 @@ const api = {
...
+    try {
+        response = await fetcher()
+    } catch (e) {
+        error = e
+    }
+
+    apiStatusLogic.findMounted()?.actions.onApiResponse(response, error)
+
+    if (error || !response) {
+        throw { status: 0, message: error ?? 'Unknown error' }
```
Comment (pauldambra) [inline frontend/src/lib/api.ts:2106]: really we should throw an actual error class here (and below)

File: frontend/src/lib/api.ts [hunk 2]
```diff
@@ -2119,10 +2089,36 @@ const api = {
...
+    }
+
+    if (!response.ok) {
+        const duration = new Date().getTime() - startTime
+        const pathname = new URL(url, location.origin).pathname
+        posthog.capture('client_request_failure', { pathname, method, duration, status: response.status })
+
+        const data = await getJSONOrThrow(response)
```
Comment (pauldambra) [inline frontend/src/lib/api.ts:2114]: which reminds me that `getJSONOrThrow` doesn't throw

File: frontend/src/lib/logic/apiStatusLogic.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                setInternetConnectionIssue: (_, { issue }) => issue,
+            },
+        ],
+    }),
+    listeners(({ cache, actions, values }) => ({
+        onApiResponse: async ({ response, error }, breakpoint) => {
+            if (error || !response?.status) {
+                // Likely CORS headers errors (i.e. request failing without reaching Django))
+                if ((error as any)?.message === 'Failed to fetch') {
+                    actions.setInternetConnectionIssue(true)
```
Comment (pauldambra) [inline frontend/src/lib/logic/apiStatusLogic.ts:25]: i wonder if we should breakpoint(50) here to avoid flapping this status?

File: frontend/__snapshots__/scenes-app-sidepanels--side-panel-settings--dark.png
Comment (pauldambra) [inline frontend/__snapshots__/scenes-app-sidepanels--side-panel-settings--dark.png:1]: these all no longer have quick start in the side panel 🤷

**Review by pauldambra — APPROVED** [review]
didn't run it, but LGTM

File: frontend/src/lib/api.ts (hunk 1 above)
Comment (benjackwhite) [reply frontend/src/lib/api.ts:2106]: good idea

File: frontend/src/lib/logic/apiStatusLogic.ts (hunk 3 above)
Comment (benjackwhite) [reply frontend/src/lib/logic/apiStatusLogic.ts:25]: Good thinking!

File: frontend/__snapshots__/scenes-app-sidepanels--side-panel-settings--dark.png
Comment (benjackwhite) [reply frontend/__snapshots__/scenes-app-sidepanels--side-panel-settings--dark.png:1]: Yeah this is correct but I have no idea why the snapshots are only now getting updated

File: frontend/src/lib/api.ts [hunk 4]
```diff
@@ -2119,10 +2096,33 @@ const api = {
...
+
+    let response
+    let error
+    try {
+        response = await fetcher()
+    } catch (e) {
+        error = e
+    }
+
+    apiStatusLogic.findMounted()?.actions.onApiResponse(response, error)
```
Comment (Twixes) [inline frontend/src/lib/api.ts:2110]: I suppose this should `always` be mounted? In that case might be worth doing `useMountedLogic(apiStatusLogic)` in the root `App` component, instead of tying to `sceneLogic`

File: frontend/src/layout/navigation/ProjectNotice.tsx [hunk 5]
```diff
@@ -108,6 +108,15 @@ export function ProjectNotice(): JSX.Element | null {
                 children: 'Log out',
             },
         },
+        internet_connection_issue: {
+            message: 'PostHog is having trouble connecting to the internet. Please check your connection.',
```
Comment (Twixes) [inline frontend/src/layout/navigation/ProjectNotice.tsx:112]: Maybe "connecting to the server"? There is also a low chance that we're down or an endpoint is acting up

File: frontend/src/lib/logic/apiStatusLogic.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                if (now - 10000 > (cache.lastUnauthorizedCheck ?? 0)) {
+                    cache.lastUnauthorizedCheck = Date.now()
+
+                    await breakpoint(100)
+
+                    await api.get('api/users/@me/').catch((error: any) => {
+                        if (error.status === 401) {
+                            window.location.href = '/logout'
+                        }
+                    })
```
Comment (Twixes) [inline frontend/src/lib/logic/apiStatusLogic.ts:48]: Weirdly I've not been able to get this to work. Can we have a Cypress test for this logic? There are a lot of building blocks here, so I think E2E would be useful, and at the same time the intended behavior is straightforward so should be easy to test (`fetch('/logout')`, then any click should redirect to login)

**Review by Twixes — COMMENTED** [review]
Looks good, but I'm not sure auto-logout works – may be just me though, so just flagging for extra confirmation

File: frontend/src/lib/logic/apiStatusLogic.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+            if (response?.status === 401) {
+                // api.ts calls this if we see a 401
+                const now = Date.now()
+
+                // We don't want to check too often (multiple api calls might fail at once)
+                if (now - 10000 > (cache.lastUnauthorizedCheck ?? 0)) {
+                    cache.lastUnauthorizedCheck = Date.now()
+
+                    await breakpoint(100)
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/logic/apiStatusLogic.ts:42]: Hmm i guess its covered by the check above it...

File: frontend/src/lib/logic/apiStatusLogic.ts (hunk 6 above)
Comment (benjackwhite) [reply frontend/src/lib/logic/apiStatusLogic.ts:46]: It works perfectly for me. Two tabs logged in. Log out of 1 and then click something in the other and it will trigger this flow. It's hard to test in a mocked way as there is other routing that gets in the way.

I'm not against a cypress test - I just have no idea how one would simulate that!

Comment (Twixes) [reply frontend/src/lib/logic/apiStatusLogic.ts:46]: Looks like I just had some unknown local problem, added a test (take a look to verify) and now the logic works perfectly (confirmed the test only passes on this branch and fails on master)

**Review by Twixes — APPROVED** [review]
Looks good

File: frontend/src/layout/navigation/ProjectNotice.tsx [hunk 8]
```diff
@@ -108,6 +108,15 @@ export function ProjectNotice(): JSX.Element | null {
                 children: 'Log out',
             },
         },
+        internet_connection_issue: {
+            message: 'PostHog is having trouble connecting to the server. Please check your connection.',
```
Comment (raquelmsmith) [inline frontend/src/layout/navigation/ProjectNotice.tsx:112]: This sounds at first read like _PostHog_ has a problem. I'd rephrase like:
```suggestion
            message: 'You're offline. Please check your connection.',
```
